lib/xandra/cluster/control_connection.ex

```diff
@@ -0,0 +1,119 @@
+
+  require Logger
+
+  @default_timeout 5_000
+  @socket_options [packet: :raw, mode: :binary, active: false]
+
+  defstruct [:address, :port, :cluster, :socket, new: true, buffer: <<>>]
+
+  def start_link(cluster, address, port) do
+    state = %__MODULE__{cluster: cluster, address: address, port: port}
+    Connection.start_link(__MODULE__, state)
+  end
+
+  def init(state) do
+    {:connect, :init, state}
+  end
+
+  def connect(_action, %__MODULE__{address: address, port: port} = state) do
+    case :gen_tcp.connect(address, port, @socket_options, @default_timeout) do
+      {:ok, socket} ->
+        state = %{state | socket: socket}
+        with {:ok, supported_options} <- Utils.request_options(socket),
+             :ok <- startup_connection(socket, supported_options),
+             :ok <- register_to_events(socket),
+             :ok <- :inet.setopts(socket, active: :true),
+             {:ok, state} <- report_active(state) do
+          {:noreply, state}
+        else
+          {:error, _reason} = error ->
+            {:connect, :reconnect, state} = disconnect(error, state)
+            {:backoff, 5_000, state}
+        end
+      {:error, _reason} ->
+        {:backoff, 5_000, state}
+    end
+  end
+
+  def handle_info({:tcp_error, socket, reason}, %__MODULE__{socket: socket} = state) do
+    {:disconnect, {:error, reason}, state}
+  end
+
+  def handle_info({:tcp_closed, socket}, %__MODULE__{socket: socket} = state) do
+    {:disconnect, {:error, :closed}, state}
+  end
+
+  def handle_info({:tcp, socket, data}, %__MODULE__{socket: socket} = state) do
+    state = %{state | buffer: state.buffer <> data}
+    {:noreply, report_event(state)}
+  end
+
+  def disconnect({:error, _reason}, %__MODULE__{} = state) do
+    :gen_tcp.close(state.socket)
+    {:connect, :reconnect, %{state | socket: nil, buffer: <<>>}}
+  end
+
+  defp report_active(%{new: false} = state) do
+    {:ok, state}
+  end
+
+  defp report_active(%{new: true, cluster: cluster, socket: socket} = state) do
+    with {:ok, {address, port}} <- :inet.peername(socket) do
+      Xandra.Cluster.start_pool(cluster, address, port)
+      {:ok, %{state | new: false, address: address}}
+    end
+  end
+
+  defp startup_connection(socket, supported_options) do
+    %{"CQL_VERSION" => [cql_version | _]} = supported_options
+    requested_options = %{"CQL_VERSION" => cql_version}
+    Utils.startup_connection(socket, requested_options)
+  end
+
+  defp register_to_events(socket) do
+    payload =
+      Frame.new(:register)
+      |> Protocol.encode_request(["STATUS_CHANGE"])
+      |> Frame.encode()
+
+    with :ok <- :gen_tcp.send(socket, payload),
+         {:ok, %Frame{} = frame} <- Utils.recv_frame(socket) do
+      :ok = Protocol.decode_response(frame)
+    else
+      {:error, reason} ->
+        {:error, reason}
+    end
+  end
+
+  defp report_event(%{cluster: cluster, buffer: buffer} = state) do
+    case decode_frame(buffer) do
```

Comment:
`decode_frame/1` only decodes the next frame if a full frame is available. This causes a problem when more than one C* frame ends up in a single TCP frame. In that case, we'll just parse the next frame, and put the `rest` in the buffer, but that `rest` could contain one or more complete frames still. We should always recurse and only stop recursing when we finally get `:error` (most of the time, it will be because we have no more data to consume, because one TCP packet was sent for one C* frame):

```elixir
def report_event(state) do
  case decode_frame(buffer) do
    {frame, rest} ->
      ...
      report_event(%{state | buffer: rest})
    :error ->
      state
  end
end
```